vortex-array/src/arrays/primitive/compute/take/avx2.rs

```diff
@@ -0,0 +1,350 @@
+use num_traits::AsPrimitive;
+use vortex_buffer::{Alignment, Buffer, BufferMut};
+use vortex_dtype::{
+    NativePType, PType, match_each_integer_ptype, match_each_native_ptype,
+    match_each_unsigned_integer_ptype,
+};
+use vortex_error::{VortexResult, vortex_panic};
+
+use crate::arrays::primitive::PrimitiveArray;
+use crate::arrays::primitive::compute::take::{TakeImpl, take_primitive_scalar};
+use crate::validity::Validity;
+use crate::{ArrayRef, IntoArray};
+
+pub(super) struct TakeKernelAVX2;
+
+impl TakeImpl for TakeKernelAVX2 {
+    #[allow(clippy::cognitive_complexity)]
+    fn take(
+        &self,
+        values: &PrimitiveArray,
+        indices: &PrimitiveArray,
+        validity: Validity,
+    ) -> VortexResult<ArrayRef> {
+        if values.ptype() != PType::F16
+            && indices.dtype().is_unsigned_int()
+            && indices.all_valid()?
+            && values.all_valid()?
+        {
+            match_each_unsigned_integer_ptype!(indices.ptype(), |I| {
+                match_each_native_ptype!(values.ptype(), |V| {
+                    Ok(take_primitive_avx2(
+                        indices.as_slice::<I>(),
+                        values.as_slice::<V>(),
+                        validity,
+                    )
+                    .into_array())
+                })
+            })
+        } else {
+            match_each_native_ptype!(values.ptype(), |T| {
+                match_each_integer_ptype!(indices.ptype(), |I| {
+                    // NOTE: we don't need to pre-scan the indices here, because
+                    //  the slice indexing inside of take_primitive_scalar will do bounds checks.
+                    let result =
+                        take_primitive_scalar(values.as_slice::<T>(), indices.as_slice::<I>());
+                    Ok(PrimitiveArray::new(result, validity).into_array())
+                })
+            })
+        }
+    }
+}
+
+/// The main gather function that is used by the inner loop kernel for AVX2 gather.
+pub(crate) trait GatherFn<Idx, Values> {
+    /// The number of data elements that are written to the `dst` on each loop iteration.
+    const WIDTH: usize;
+    /// The number of indices read from `indices` on each loop iteration.
+    /// Depending on the available instructions and bit-width we may stride by a larger amount
+    /// than we actually end up reading from `src` (governed by the `WIDTH` parameter).
+    const STRIDE: usize = Self::WIDTH;
+
+    /// Gather values from `src` into the `dst` using the `indices`, optionally using
+    /// SIMD instructions.
+    ///
+    /// # Safety
+    ///
+    /// This function can read up to `STRIDE` elements through `indices`, and read/write up to
+    /// `WIDTH` elements through `src` and `dst` respectively.
+    unsafe fn gather(indices: *const Idx, src: *const Values, dst: *mut Values);
+}
+
+/// AVX2 version of GatherFn defined for 32- and 64-bit value types.
+enum AVX2Gather {}
+
+unsafe fn identity<T>(input: T) -> T {
```

Comment:
some of the impls don't need an `extend` operation (when the indices and values are the same size) so we use this and it should get optimized away